File: ruby/snakesladders/lib/snakesladders/models/board.rb
module Snakesladders
  module Models
    class Board
      attr_accessor :size, :snake_points, :ladder_points, :all_points

      def initialize(size, snakes, ladders)
        self.size = size
        self.all_points = {}
        self.snake_points = {}
        self.ladder_points = {}

        mark_snakes(snakes)
        mark_ladders(ladders)
      end

      def mark_snakes(snakes)
        snakes.each do |snake|
          if all_points[snake.start_point] || all_points[snake.end_point]
            raise ::Snakesladders::Errors::OverlappingPointError.new("Snake overlaps with existing snake or ladder #{snake.start_point}, #{snake.end_point}")
          end

          raise ::Snakesladders::Errors::BoardOutOfBoundError.new("Snake start_point: #{snake.start_point} is outside board: #{last_cell}") if snake.start_point > last_cell

          all_points[snake.start_point] = 1
          all_points[snake.end_point] = 1
          snake_points[snake.start_point] = snake.end_point
        end
      end

      def mark_ladders(ladders)
        ladders.each do |ladder|
          if all_points[ladder.start_point] || all_points[ladder.end_point]
            raise ::Snakesladders::Errors::OverlappingPointError.new("ladder overlaps with existing ladder or ladder #{ladder.start_point}, #{ladder.end_point}")
          end

          raise ::Snakesladders::Errors::BoardOutOfBoundError.new("ladder start_point: #{ladder.end_point} is outside board: #{last_cell}") if ladder.end_point > last_cell

          all_points[ladder.start_point] = 1
          all_points[ladder.end_point] = 1
          ladder_points[ladder.start_point] = ladder.end_point
        end
      end

      def last_cell
        size * size
      end
    end
  end
end

File: java/snakes-and-ladders/src/main/java/gagan93jtg/lld/models/Board.java
package gagan93jtg.lld.models;

import gagan93jtg.lld.errors.ObjectOutOfRange;

import java.security.InvalidParameterException;
import java.util.ArrayList;
import java.util.HashMap;

public class Board {
  private int size;
  private HashMap<Integer, Integer> snakePoints, ladderPoints, allPoints;

  public Board(int size, ArrayList<Snake> snakes, ArrayList<Ladder> ladders) {
    this.size = size;
    this.allPoints = new HashMap<>();

    markSnakes(snakes);
    markLadders(ladders);
  }

  public int getLastCell() {
    return size * size;
  }

  public HashMap<Integer, Integer> getSnakePoints() {
    return snakePoints;
  }

  public HashMap<Integer, Integer> getLadderPoints() {
    return ladderPoints;
  }

  private void markSnakes(ArrayList<Snake> snakes) {
    this.snakePoints = new HashMap<>();
    for (Snake snake : snakes) {
      int snakeStartValue = allPoints.putIfAbsent(snake.getStart(), 1) == null ? 0 : 1;
      int snakeEndValue = allPoints.putIfAbsent(snake.getEnd(), 1) == null ? 0 : 1;

      if (snakeStartValue == 1 || snakeEndValue == 1) {
        throw new InvalidParameterException("Snake overlaps with existing snake or ladder " + snake.getStart() + ", " + snake.getEnd());
      }

      if (snake.getStart() > getLastCell()) {
        throw new ObjectOutOfRange("Snake start value " + snake.getStart() + " is out of board range " + getLastCell());
      }
      snakePoints.put(snake.getStart(), snake.getEnd());
    }
  }

  private void markLadders(ArrayList<Ladder> ladders) {
    this.ladderPoints = new HashMap<>();
    for (Ladder ladder : ladders) {
      int ladderStartValue = allPoints.putIfAbsent(ladder.getStart(), 1) == null ? 0 : 1;
      int ladderEndValue = allPoints.putIfAbsent(ladder.getEnd(), 1) == null ? 0 : 1;

      if (ladderStartValue == 1 || ladderEndValue == 1) {
        throw new InvalidParameterException("Ladder overlaps with existing ladder or snake " + ladder.getStart() + ", " + ladder.getEnd());
      }

      if (ladder.getEnd() > getLastCell()) {
        throw new ObjectOutOfRange("Ladder start value " + ladder.getStart() + " is out of board range " + getLastCell());
      }
      ladderPoints.put(ladder.getStart(), ladder.getEnd());
    }
  }
}
